lucet-wasi/src/hostcalls/fs.rs

```diff
@@ -0,0 +1,1338 @@
+            Ok(filestat) => {
+                let host_filestat = host::filestat_from_nix(filestat);
+                unsafe {
+                    enc_filestat_byref(vmctx, filestat_ptr, host_filestat)
+                        .expect("can write into the pointer");
+                }
+            }
+        },
+        Err(e) => return enc_errno(e),
+    }
+    wasm32::__WASI_ESUCCESS
+}
+
+pub fn wasi_path_filestat_get(
+    vmctx: &mut Vmctx,
+    dirfd: wasm32::__wasi_fd_t,
+    dirflags: wasm32::__wasi_lookupflags_t,
+    path_ptr: wasm32::uintptr_t,
+    path_len: wasm32::size_t,
+    filestat_ptr: wasm32::uintptr_t,
+) -> wasm32::__wasi_errno_t {
+    use nix::fcntl::AtFlags;
+    use nix::sys::stat::fstatat;
+
+    let dirfd = dec_fd(dirfd);
+    let dirflags = dec_lookupflags(dirflags);
+    let path = match unsafe { dec_slice_of::<u8>(vmctx, path_ptr, path_len) } {
+        Ok((ptr, len)) => OsStr::from_bytes(unsafe { std::slice::from_raw_parts(ptr, len) }),
+        Err(e) => return enc_errno(e),
+    };
+    let rights = host::__WASI_RIGHT_PATH_FILESTAT_GET;
+    let (dir, path) = match path_get(&vmctx, dirfd, dirflags, path, rights.into(), 0, false) {
+        Ok((dir, path)) => (dir, path),
+        Err(e) => return enc_errno(e),
+    };
+    let atflags = match dirflags {
+        wasm32::__WASI_LOOKUP_SYMLINK_FOLLOW => AtFlags::empty(),
+        _ => AtFlags::AT_SYMLINK_NOFOLLOW,
+    };
+    match fstatat(dir, path.as_os_str(), atflags) {
+        Err(e) => wasm32::errno_from_nix(e.as_errno().unwrap()),
+        Ok(filestat) => {
+            let host_filestat = host::filestat_from_nix(filestat);
+            unsafe {
+                enc_filestat_byref(vmctx, filestat_ptr, host_filestat)
+                    .expect("can write into the pointer");
+            }
+            wasm32::__WASI_ESUCCESS
+        }
+    }
+}
+
+pub fn wasi_path_create_directory(
+    vmctx: &mut Vmctx,
+    dirfd: wasm32::__wasi_fd_t,
+    path_ptr: wasm32::uintptr_t,
+    path_len: wasm32::size_t,
+) -> wasm32::__wasi_errno_t {
+    use nix::errno;
+    use nix::libc::mkdirat;
+
+    let dirfd = dec_fd(dirfd);
+    let path = match unsafe { dec_slice_of::<u8>(vmctx, path_ptr, path_len) } {
+        Ok((ptr, len)) => OsStr::from_bytes(unsafe { std::slice::from_raw_parts(ptr, len) }),
+        Err(e) => return enc_errno(e),
+    };
+    let rights = host::__WASI_RIGHT_PATH_CREATE_DIRECTORY;
+    let (dir, path) = match path_get(&vmctx, dirfd, 0, path, rights.into(), 0, false) {
+        Ok((dir, path)) => (dir, path),
+        Err(e) => return enc_errno(e),
+    };
+    let path_cstr = match std::ffi::CString::new(path.as_os_str().as_bytes()) {
+        Ok(path_cstr) => path_cstr,
+        Err(_) => return wasm32::__WASI_EINVAL,
+    };
+    // nix doesn't expose mkdirat() yet
+    match unsafe { mkdirat(dir, path_cstr.as_ptr(), 0o777) } {
+        0 => wasm32::__WASI_ESUCCESS,
+        _ => wasm32::errno_from_nix(errno::Errno::last()),
```

Comment:
It looks like `last()` translates into `ENOSYS`, so it'd be a little more obvious here if this (and other places that do this) just said `wasm32::errno_from_nix(errno::Errnor::ENOSYS)` directly.

Reply:
Does it? `nix::errno::Errno::last()` is documented as being the last `errno` value, and `mkdirat()` can return many more error codes than `ENOSYS`.

Reply:
I mean that `errno_from_nix` turns `nix::errno::Errno::last()` into `wasm32::__ENOSYS`, so `wasm32::errno_from_nix(errno::Errnor::ENOSYS)` should produce the same result.

Reply:
Hey Dan.

I'm still a little bit confused by your comment.

[`nix::errno::Errno::last()`](https://github.com/nix-rust/nix/blob/63aec5066b6bca637f45a4acc23aa8f8079ca050/src/errno.rs#L132-L134) returns the current value of libc's `errno`.

And `mkdirat()` can fail with `EEXIST` that `wasm32::errno_from_nix()` turns into `__WASI_EEXIST`.  I added a trivial [test for this](https://github.com/fastly/lucet/blob/4321e538acd868e9303a9576a5138b64b43542ef/lucet-wasi/tests/guests/fs.c#L88) and this doesn't seem to return `ENOSYS`, but `EEXIST` as expected.

Can you clarify when `errno_from_nix` turns `nix::errno::Errno::last()` into `wasm32::__ENOSYS`?

Reply:
Ah, I somehow missed that `nix::errno::Errno::last()` returns the current `errno` value. You're right, and the code looks good!

Reply:
👍